File: frontend/src/types/inventory.ts
import { z } from "zod";

export const SiteSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const PlacementDomainSchema = z.object({
  id: z.string(),
  name: z.string(),
  kind: z.string().transform((v) => (v === "shared" ? "shared" : "site")),
  description: z.string().nullable(),
  icon: z.string().nullable(),
  site_id: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const LoadBalancerSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  ip_address: z.string().nullable(),
  site_id: z.string().nullable(),
  is_local: z.boolean(),
  created_at: z.string(),
  updated_at: z.string(),
});

export type Site = z.infer<typeof SiteSchema>;
export type PlacementDomainRecord = z.infer<typeof PlacementDomainSchema>;
export type LoadBalancer = z.infer<typeof LoadBalancerSchema>;

export type SiteCreatePayload = { name: string; description?: string | null };
export type SiteUpdatePayload = { name?: string; description?: string | null };

export type PlacementDomainCreatePayload = {
  name: string;
  kind?: "site" | "shared";
  description?: string | null;
  icon?: "site" | "shared" | "building" | null;
  site_id?: string | null;
};

export type PlacementDomainUpdatePayload = {
  name?: string;
  kind?: "site" | "shared";
  description?: string | null;
  icon?: "site" | "shared" | "building" | null;
  site_id?: string | null;
};

export type LoadBalancerCreatePayload = {
  name: string;
  description?: string | null;
  ip_address?: string | null;
  site_id?: string | null;
};

export type LoadBalancerUpdatePayload = {
  name?: string;
  description?: string | null;
  ip_address?: string | null;
  site_id?: string | null;
};

File: backend/app/schemas/inventory.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

PlacementKind = Literal["site", "shared"]
PlacementIcon = Literal["site", "shared", "building"]


class SiteCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)


class SiteUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)


class SiteRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str | None
    created_at: datetime
    updated_at: datetime


class PlacementDomainCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    kind: PlacementKind = "site"
    description: str | None = Field(default=None, max_length=2000)
    icon: PlacementIcon | None = None
    site_id: str | None = None


class PlacementDomainUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    kind: PlacementKind | None = None
    description: str | None = Field(default=None, max_length=2000)
    icon: PlacementIcon | None = None
    site_id: str | None = None


class PlacementDomainRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    kind: str
    description: str | None
    icon: str | None
    site_id: str | None
    created_at: datetime
    updated_at: datetime


class LoadBalancerCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)
    ip_address: str | None = Field(default=None, max_length=64)
    site_id: str | None = None
    is_local: bool = False


class LoadBalancerUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=2000)
    ip_address: str | None = Field(default=None, max_length=64)
    site_id: str | None = None


class LoadBalancerRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    description: str | None
    ip_address: str | None
    site_id: str | None
    is_local: bool
    created_at: datetime
    updated_at: datetime
